c10/core/DispatchKeySet.h

```diff
@@ -174,18 +443,41 @@ class DispatchKeySet final {
     }
 
     bool operator==(const self_type& rhs) const {
-      return i_ == rhs.i_;
+      return functionality_mask_ == rhs.functionality_mask_ &&
+             functionality_idx_ == rhs.functionality_idx_ &&
+             backend_mask_ == rhs.backend_mask_ &&
+             backend_idx_ == rhs.backend_idx_;
     }
     bool operator!=(const self_type& rhs) const {
-      return i_ != rhs.i_;
+      return functionality_mask_ != rhs.functionality_mask_ ||
+             functionality_idx_ != rhs.functionality_idx_ ||
+             backend_mask_ != rhs.backend_mask_ ||
+             backend_idx_ != rhs.backend_idx_;
     }
     DispatchKey operator*() const {
-      return static_cast<DispatchKey>(i_);
+      auto functionality_key = static_cast<DispatchKey>(functionality_idx_);
+      if (isPerBackendFunctionalityKey(functionality_key)) {
+         auto next_key = toRuntimePerBackendFunctionalityKey(functionality_key, static_cast<BackendComponent>(backend_idx_));
+         // We expect all of the Dense, Sparse, Quantized, and Autograd keys to be ordered the same way
+         // with respect to their backends
+         TORCH_INTERNAL_ASSERT(toBackendComponent(next_key) == static_cast<BackendComponent>(backend_idx_),
+           "Tried to map functionality key ", toString(functionality_key), " and backend bit ",
+           toString(static_cast<BackendComponent>(backend_idx_)), " to a runtime key, but ended up with ", toString(next_key),
+           ". This can happen if the order of the backend dispatch keys in DispatchKey.h isn't consistent.",
+           " Please double check that enum for inconsistencies.");
+         return next_key;
+      } else {
+        return functionality_key;
+      }
     }
 
    private:
     const uint64_t* data_ptr_;
-    uint8_t i_;
+    uint8_t functionality_mask_;
+    uint8_t backend_mask_;
+    uint8_t functionality_idx_;
```

Comment:
These two `idx_` objects are only needed for the `operator*` functions as the mask get reset (when reaching last backend for example). I think that this could be reflected in the name?

Also these two objects index into `DispatchKey` and `BackendComponent` (and not `DispatchKeySet.repr_` like the two above).

Maybe `current_dispatchkey_idx_` and `current_backendcomponent_idx_`?
And make the two above: `next_functionality_` and `next_backend_`?